### PR TITLE
Reorganize candidate rewrite rule filtering

### PR DESCRIPTION
**Pull request by ajreynol**
This refactors how candidate rewrite rules are filtered.

Previously, we had a class "SygusSamplerExt" which inherited from the base sampler class "SygusSampler", and acted as both a sampler and a rewrite rule filter.

Now, "SygusSamplerExt" has been renamed "CandidateRewriteFilter" in a new file, does not inherit from SygusSampler, has the sole purpose of being a rewrite rule filter. CandidateRewriteDatabase is updated to use a normal SygusSampler + CandidateRewriteFilter as two separate objects.

The majority of this PR is a code move.

### CODE REVIEW

File: src/theory/quantifiers/candidate_rewrite_filter.cpp [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+bool CandidateRewriteFilter::filterPair(Node n, Node eq_n)
+{
+  Node bn = n;
+  Node beq_n = eq_n;
+  if (d_use_sygus_type)
+  {
+    bn = d_tds->sygusToBuiltin(n);
+    beq_n = d_tds->sygusToBuiltin(eq_n);
+  }
+  Trace("sygus-synth-rr") << "sygusSampleExt : " << bn << "..." << beq_n
```
Comment (4tXJ7f) [inline src/theory/quantifiers/candidate_rewrite_filter.cpp]: sygusSampleExt -> CandidateRewriteFilter

File: src/theory/quantifiers/candidate_rewrite_filter.cpp [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+
+bool CandidateRewriteFilter::notify(Node s,
+                                    Node n,
+                                    std::vector<Node>& vars,
+                                    std::vector<Node>& subs)
+{
+  Assert(!d_curr_pair_rhs.isNull());
+  std::map<Node, std::unordered_set<Node, NodeHashFunction> >::iterator it =
+      d_pairs.find(n);
+  if (Trace.isOn("sse-match"))
```
Comment (4tXJ7f) [inline src/theory/quantifiers/candidate_rewrite_filter.cpp]: sse-match -> cr-filter (or something similar)

File: src/theory/quantifiers/candidate_rewrite_database.h [hunk 3]
```diff
@@ -116,11 +117,9 @@ class CandidateRewriteDatabase
...
    * rewrite rules.
    */
-  SygusSamplerExt d_sampler;
-  /** a (dummy) user context, used for d_drewrite */
-  context::UserContext d_fake_context;
-  /** dynamic rewriter class */
-  std::unique_ptr<DynamicRewriter> d_drewrite;
+  SygusSampler d_sampler;
+  /** candidate rewrite filter */
+  CandidateRewriteFilter d_crewrite_filter;
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/quantifiers/candidate_rewrite_database.h]: Going to keep this for now for consistency within this class.

[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/theory/quantifiers/candidate_rewrite_database.h]: Ok.

**Pull request merged**